server/src/infra/instrumentation.ts

```diff
@@ -0,0 +1,93 @@
+import { serverVersion } from '@app/domain/domain.constant';
+import { Histogram, MetricOptions, ValueType, metrics } from '@opentelemetry/api';
+import { AsyncLocalStorageContextManager } from '@opentelemetry/context-async-hooks';
+import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
+import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+import { IORedisInstrumentation } from '@opentelemetry/instrumentation-ioredis';
+import { NestInstrumentation } from '@opentelemetry/instrumentation-nestjs-core';
+import { PgInstrumentation } from '@opentelemetry/instrumentation-pg';
+import { Resource } from '@opentelemetry/resources';
+import { ExplicitBucketHistogramAggregation, View } from '@opentelemetry/sdk-metrics';
+import { NodeSDK } from '@opentelemetry/sdk-node';
+import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
+import { snakeCase, startCase } from 'lodash';
+import { OpenTelemetryModuleOptions } from 'nestjs-otel/lib/interfaces';
+import { copyMetadataFromFunctionToFunction } from 'nestjs-otel/lib/opentelemetry.utils';
+import { performance } from 'node:perf_hooks';
+import { excludePaths } from './infra.config';
+import { DecorateAll } from './infra.utils';
+
+const aggregation = new ExplicitBucketHistogramAggregation(
+  [0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000],
+  true,
+);
+
+export const otelSDK = new NodeSDK({
+  resource: new Resource({
+    [SemanticResourceAttributes.SERVICE_NAME]: `immich`,
+    [SemanticResourceAttributes.SERVICE_VERSION]: serverVersion.toString(),
+  }),
+  metricReader: new PrometheusExporter({ port: 8081 }),
+  contextManager: new AsyncLocalStorageContextManager(),
+  instrumentations: [
+    new HttpInstrumentation(),
+    new IORedisInstrumentation(),
+    new NestInstrumentation(),
+    new PgInstrumentation(),
+  ],
+  views: [new View({ aggregation, instrumentName: '*', instrumentUnit: 'ms' })],
+});
+
+export const otelConfig: OpenTelemetryModuleOptions = {
+  metrics: {
+    hostMetrics: process.env.OTEL_SDK_DISABLED !== 'true',
+    apiMetrics: {
+      enable: process.env.OTEL_SDK_DISABLED !== 'true',
+      ignoreRoutes: excludePaths,
+    },
+  },
```

Comment:
I think we should use an env to enable this functionality not the other way around.

Also we should probably make the port configurable. IMMICH_METRICS_PORT or something.

Reply:
> I think we should use an env to enable this functionality not the other way around.

That's how it currently works. This env is set to false in the Dockerfile so you have to set it yourself to enable it.

I agree with the point about customizing the port. It might also be good to allow only enabling e.g. api metrics without enabling everything.

Reply:
Hmm, that's a good point. I didn't think about bare-metal deployments. Changing the comparison to `=== false` would do the trick, but this will also get handled after I add envs for host, api and repo metrics.

Reply:
I see that it is an external variable that may have other implications. I wonder if it is possible to just check "does the disabled env exist"? If it doesn't then default it to disabled. Basically do the "defaulting" in the code instead of the Dockerfile.

Reply:
I wasn't sure about when the env gets checked so I looked and it's when NodeSDK gets constructed. I'm thinking we can check the metric envs and set 'OTEL_SDK_DISABLED=true' before making the sdk if it's unset and none of them are enabled.

Reply:
Yeah, that sounds perfect.